File: framework/languages/dotnet/samples/Bingo/Server/Play/BingoRoomSpots/Handlers/SubmitBingoCardHandler.cs
using Bingo.Server.Play.Actors;
using Bingo.Shared.Contracts;
using Zlink.Framework.Contracts.Spots;

namespace Bingo.Server.Play.BingoRoomSpots.Handlers;

internal sealed class SubmitBingoCardHandler
    : IZLinkSpotActorRequestHandler<BingoRoomSpot, PlayerActor, SubmitBingoCardReq, SubmitBingoCardRes>
{
    public async ValueTask<SubmitBingoCardRes> HandleAsync(
        BingoRoomSpot spot,
        PlayerActor actor,
        ZLinkSpotActorRequestContext context,
        SubmitBingoCardReq message,
        CancellationToken cancellationToken)
    {
        _ = context;
        spot.EnsureRoomId(message.RoomId);
        var card = BingoCard.FromSubmittedNumbers(message.Card);
        var change = spot.SubmitCard(actor.ActorId, card);
        if (change.ShouldStartDrawTimer)
        {
            await spot.StartDrawTimerAsync(cancellationToken);
        }

        return new SubmitBingoCardRes(change.State);
    }
}

File: framework/languages/java/samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/infrastructure/zlink/spots/bingoroomspot/handlers/SubmitBingoCardHandler.kt
package systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.handlers

import systems.zlink.framework.kotlin.ZLinkSuspendingSpotActorRequestHandler
import systems.zlink.framework.spots.ZLinkSpotActorRequestContext
import systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.actors.PlayerActor
import systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.BingoRoomSpot
import systems.zlink.samples.kotlin.bingo.shared.contracts.SubmitBingoCardReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.SubmitBingoCardRes

class SubmitBingoCardHandler() : ZLinkSuspendingSpotActorRequestHandler<
    BingoRoomSpot,
    PlayerActor,
    SubmitBingoCardReq,
    SubmitBingoCardRes,
    > {
    override suspend fun handle(
        spot: BingoRoomSpot,
        actor: PlayerActor,
        context: ZLinkSpotActorRequestContext,
        request: SubmitBingoCardReq,
    ): SubmitBingoCardRes =
        spot.submitCard(actor, request)
}
